url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/de/odysseus/el/misc/NumberOperations.java

```diff
@@ -0,0 +1,179 @@
+		}
+		return converter.convert(o1, Long.class) + converter.convert(o2, Long.class);
+	}
+
+	public static final Number sub(TypeConverter converter, Object o1, Object o2) {
+		if (o1 == null && o2 == null) {
+			return LONG_ZERO;
+		}
+		if (o1 instanceof BigDecimal || o2 instanceof BigDecimal) {
+			return converter.convert(o1, BigDecimal.class).subtract(converter.convert(o2, BigDecimal.class));
+		}
+		if (isFloatOrDoubleOrDotEe(o1) || isFloatOrDoubleOrDotEe(o2)) {
+			if (o1 instanceof BigInteger || o2 instanceof BigInteger) {
+				return converter.convert(o1, BigDecimal.class).subtract(converter.convert(o2, BigDecimal.class));
+			}
+			return converter.convert(o1, Double.class) - converter.convert(o2, Double.class);
+		}
+		if (o1 instanceof BigInteger || o2 instanceof BigInteger) {
+			return converter.convert(o1, BigInteger.class).subtract(converter.convert(o2, BigInteger.class));
+		}
+		return converter.convert(o1, Long.class) - converter.convert(o2, Long.class);
+	}
+
+	public static final Number mul(TypeConverter converter, Object o1, Object o2) {
+		if (o1 == null && o2 == null) {
+			return LONG_ZERO;
+		}
+		if (o1 instanceof BigDecimal || o2 instanceof BigDecimal) {
+			return converter.convert(o1, BigDecimal.class).multiply(converter.convert(o2, BigDecimal.class));
+		}
+		if (isFloatOrDoubleOrDotEe(o1) || isFloatOrDoubleOrDotEe(o2)) {
+			if (o1 instanceof BigInteger || o2 instanceof BigInteger) {
+				return converter.convert(o1, BigDecimal.class).multiply(converter.convert(o2, BigDecimal.class));
+			}
+			return converter.convert(o1, Double.class) * converter.convert(o2, Double.class);
+		}
+		if (o1 instanceof BigInteger || o2 instanceof BigInteger) {
+			return converter.convert(o1, BigInteger.class).multiply(converter.convert(o2, BigInteger.class));
+		}
+		return converter.convert(o1, Long.class) * converter.convert(o2, Long.class);
+	}
+
+	public static final Number div(TypeConverter converter, Object o1, Object o2) {
+		if (o1 == null && o2 == null) {
+			return LONG_ZERO;
+		}
+		if (isBigDecimalOrBigInteger(o1) || isBigDecimalOrBigInteger(o2)) {
+			return converter.convert(o1, BigDecimal.class).divide(converter.convert(o2, BigDecimal.class), BigDecimal.ROUND_HALF_UP);
+		}
+		return converter.convert(o1, Double.class) / converter.convert(o2, Double.class);
+	}
+
+	public static final Number mod(TypeConverter converter, Object o1, Object o2) {
+		if (o1 == null && o2 == null) {
+			return LONG_ZERO;
+		}
+		if (isBigDecimalOrFloatOrDoubleOrDotEe(o1) || isBigDecimalOrFloatOrDoubleOrDotEe(o2)) {
+			return converter.convert(o1, Double.class) % converter.convert(o2, Double.class);
+		}
+		if (o1 instanceof BigInteger || o2 instanceof BigInteger) {
+			return converter.convert(o1, BigInteger.class).remainder(converter.convert(o2, BigInteger.class));
+		}
+		return converter.convert(o1, Long.class) % converter.convert(o2, Long.class);
+	}
+
+	public static final Number neg(TypeConverter converter, Object value) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.